### ISSUE TITLE
FR: Time and pages read today in the status bar.

### ISSUE THREAD

**Issue by christian-ho** [koreader/koreader#10457]
I would find it really nice to have this information at a glance in the status bar. 

What do you think? Has it been considered yet?

**Comment by offset-torque** [comment]
+1 for this

**Comment by jospalau** [comment]
You can try to add the following piece of code in the patch I told you. I am not sure if it is the best way to do it but it works:
```
...
local SQ3 = require("lua-ljsqlite3/init")
...
DAY_STATS_SYMBOL = "TRT:"
DAY_STATS = "none"
-- Uncomment one of these to have session duration replace this item in the footer
-- DAY_STATS = "page_progress"
-- DAY_STATS = "pages_left_book"
-- DAY_STATS = "time"
-- DAY_STATS = "pages_left"
-- DAY_STATS = "battery"
-- DAY_STATS = "percentage"
-- DAY_STATS = "book_time_to_read"
-- DAY_STATS = "chapter_time_to_read"
-- DAY_STATS = "frontlight"
DAY_STATS = "mem_usage"
-- DAY_STATS = "wifi_status"
-- DAY_STATS = "book_title"
-- DAY_STATS = "book_chapter"
-- DAY_STATS = "bookmark_count"
-- DAY_STATS = "chapter_progress"
-- DAY_STATS = "frontlight_warmth"
-- DAY_STATS = "custom_text"
...
getTodayBookStats = function ()
    local now_stamp = os.time()
    local now_t = os.date("*t")
    local DataStorage = require("datastorage")
    local db_location = DataStorage:getSettingsDir() .. "/statistics.sqlite3"
    local from_begin_day = now_t.hour * 3600 + now_t.min * 60 + now_t.sec
    local start_today_time = now_stamp - from_begin_day
    local conn = SQ3.open(db_location)
    local sql_stmt = [[
        SELECT count(*),
               sum(sum_duration)
        FROM    (
                     SELECT sum(duration)    AS sum_duration
                     FROM   page_stat
                     WHERE  start_time >= %d
                     GROUP  BY id_book, page
                );
    ]]
    local today_pages, today_duration = conn:rowexec(string.format(sql_stmt, start_today_time))
    conn:close()
    if today_pages == nil then
        today_pages = 0
    end
    if today_duration == nil then
        today_duration = 0
    end
    today_duration = tonumber(today_duration)
    today_pages = tonumber(today_pages)
    return today_duration, today_pages
end

ReaderFooter.textGeneratorMap[DAY_STATS] = function(footer)
    local today_duration, today_pages = getTodayBookStats()
    local user_duration_format = "letters"
    today_duration = datetime.secondsToClockDuration(user_duration_format,today_duration, true)
    return DAY_STATS_SYMBOL .. today_duration .. "(" .. today_pages .. "p)"
end
```
![IMG_20230520_200650](https://github.com/koreader/koreader/assets/85340/df09a63e-65dd-479e-a047-e20de5bbc60d)

**Comment by christian-ho** [comment]
Wow thank you it works great! :-)
Now I can do the same thing for current week/month by myself.

**Comment by christian-ho** [comment]
I was too quick ... it works but it doesn't update when turning pages.

Btw I think you could simplify your function by using this in the sql statement:
`WHERE  start_time >= strftime('%s', DATE('now', 'start of day'))`

**Comment by jospalau** [comment]
You are right. It doesn't work reliably. The minimum time by default for a page to be considered read is 5s. In any case pages are not updated reliable. It may be that the read occurs previously to the data being updated when turning page

**Comment by poire-z** [comment]
Some page read data is kept in memory in the statistics plugin, and only flushed to the sqlite db every:
https://github.com/koreader/koreader/blob/919d67656d35b7ac8cab2f581a47b74fd4ffb297/plugins/statistics.koplugin/main.lua#L31
to save SD card writes. It is also flushed when you invoke any of the statistics view, so we get update values. We (and you) surely don't want to flush it more often, like every page turn if you want that in the footer :/
So, the db you're querying may lag by ... 50 pages.

**Comment by jospalau** [comment]
I am interested more in time than in pages so for me it works even if it not accurate.

For me, does not make sense counting pages because the number varies on depending the font properties right? I normally read more than one book with different format and font sizes.

**Comment by poire-z** [comment]
> For me, does not make sense counting pages because the number varies on depending the font properties right? 

Right. But usually, one decides for the font and size he prefers reading with (at least with EPUBs), so usually mostly similar.
My time per page varies with the book difficulty or topic.

But, about this FR: getting the time and pages read today in the footer is probably quite expensive.
And "Today's timeline" #9813 is a gesture away, and will give you more accurate info, per book, + visually the periods you were reading, a lot more fun than what you could get in the footer :)

**Comment by jospalau** [comment]
That's what I use with a gesture but I still think is nice having it in the status bar.

**Comment by christian-ho** [comment]
Thanks for the comments @poire-z , that makes some things clearer!

**Comment by jospalau** [comment]
I know it is not convenient but changing the constant pointed out it works. For the moment I will use it in one of my ereaders to see if it has any impact in speed and battery. I used to be a big fan of Conky :)

**Comment by christian-ho** [comment]
Just for completeness, here is my simplified function and the SQL queries and the resulting status bar with statistics for day/week/month.

```
getTodayBookStats = function ()
    local DataStorage = require("datastorage")
    local db_location = DataStorage:getSettingsDir() .. "/statistics.sqlite3"
    local conn = SQ3.open(db_location)
    local sql_stmt = [[
        SELECT sum(duration)    AS sum_duration
        FROM   page_stat
        WHERE  start_time >= strftime('%s', DATE('now', 'start of day'));
    ]]
    local today_duration = conn:rowexec(sql_stmt)
    conn:close()
    if today_duration == nil then
        today_duration = 0
    end
    today_duration = tonumber(today_duration)
    return today_duration
end
```

Week: 
`WHERE  start_time >= strftime('%s', DATE('now', 'weekday 0', '-6 days'));`
Month: 
`WHERE  start_time >= strftime('%s', DATE('now', 'start of month'));`

![Reader_2023-06-25_084020](https://github.com/koreader/koreader/assets/57492542/8d93b8b4-43fa-4b21-ae70-081b66c0241a)
(Screen is from a few days ago)

**Comment by offset-torque** [comment]
This looks very nice. I guess this won't be implemented in KOReader but is it at least possible to add this as a **User Patch** here:

https://github.com/koreader/koreader/wiki/User-patches

**Comment by kctdfh** [comment]
This is great but anyone has any ideas on how to have the data be specific to the book you're reading right now?
Basically: time spent reading this book today, time spent reading this book this week, etc.

I looked at the sqlite schema and it's probably very simple if you have the `id_book` of the current book (you'd just have to add it to the `WHERE` clause) but for the life of me I can't figure out how to find the `id_book` from within the book itself. I assume it has to be with [getIdBookDB()](https://github.com/koreader/koreader/blob/135453776a802939d9846552bc3094baa61e80fa/plugins/statistics.koplugin/main.lua#L762) but I'm not sure if that's true and not sure how to run a plugin's function from inside of a user patch 🤔

Any help would be appreciated!

**Comment by hius07** [comment]
> how to run a plugin's function from inside of a user patch

For example
https://github.com/koreader/koreader/blob/135453776a802939d9846552bc3094baa61e80fa/frontend/ui/widget/bookmapwidget.lua#L788
